### ISSUE TITLE
Dataloader scales training batches?

### ISSUE THREAD

**Issue by PlanetMacro** [wwy155/NsDiff#7]
IIUC, NsDiff is meant to be run on raw non stationary time series data. However it appears to me as if the used dataloaders from torch_timeseries.dataloader are used in a way that applies a scaler to the raw time series:

* Used data loaders: ETTHLoader, ETTMLoader,  SlidingWindowTS

* Using SlidingWindowTS as example, it uses a scaler as input 
  https://github.com/wwy155/NsDiff/blob/main/src/experiments/prob_forecast.py#L233
  which is here 
 https://github.com/wayne155/pytorch_timeseries/blob/main/torch_timeseries/dataloader/sliding_window.py

 The scaler is applied here https://github.com/wayne155/pytorch_timeseries/blob/main/torch_timeseries/dataloader/wrapper.py#L32

------------------

Can you explain if that is the expected behavior and what the scaler is supposed to do in case of NsDiff?

**Comment by wwy155** [comment]
The training batches scale operation (x - mean)/std is a common practice in deep learning for (probabilistic) time series forecasting, for example:

1. TMDM (https://github.com/LiYuxin321/TMDM/blob/e44bde8459a0a4da749c5c1bebfb34436fca7535/TMDM/data_provider/data_loader.py#L61)
2. TimeSeriseLibrary https://github.com/thuml/Time-Series-Library/blob/947f915a338626520024c153c9c18a1c3692ad1d/data_provider/data_loader.py#L159


I have not experimented on this, but note that even with this operation, the non-stationarity is still exist. The train/test distribution shift will not be removed by a global constant operation; so NsDiff still has the advantages over other methods.

But indeed you can try to use the raw values, but the time series values ranges in [-inf, inf], leading to very unstable training and convergence in some cases (e.g. extreme values, large value disparity between variates), hence I suggest use the scalers.

**Comment by PlanetMacro** [comment]
Interesting. Reading the non-stationary transformer paper, it explicitly uses raw time series and then does the x' = (x-mean)/std internally. Importantly it afterwards scales back the result by y = y'*std + mean ....

So the NS-Transformer should use unscaled input, right? But then it also produces an unscaled f(x) for NS-Diff, so my understanding was that because of that NS-Diff should operate on the unscaled data....

Maybe one should use the NS-Transformer in  modified way to not return y'*std + mean, but y = y' to model f(x)? then we can use scaled data on NS-Diff?

Also if we use the scaler, should we then inverse it on the defusioned predictions? 

------------------------------

A bit oos, but did u tried using the Non-Stationary-Transformer itself as the diffusion model ξ_θ(., f_ϕ(.), g_ψ(.), .) ?

**Comment by wwy155** [comment]
NsDiff is a theoretical framework regardless of what mean and variance estimator you are using, the internal behaviour of the estimators can have any forms: you can even build one without training: f(x) = x_mean, g(x) = std(x - x_mean). And same for ξ_θ. Whatever f(x) is, it is just a model input x and output a tensor with size of y.

We follow TMDM to use NsTransformer, and {x' = (x-mean)/std, y = y'*std + mean} is another common practice in current time series forecasting field to handle non-stationarity for accuracy forecasting, see ReVIN[1].

Note that although ξ_θ gives us powerful inference capability, it can hardly handle distribution shift, which has to be dealt with by changing the endpoint distribution; in other words, the endpoint provides a non-stationary prior, the more accurate the prior is, the more accurate NsDiff will be. 

[1] Taesung Kim, Jinhee Kim, etc. REVERSIBLE INSTANCE NORMALIZATION FOR ACCURATE TIME-SERIES FORECASTING AGAINST DISTRIBUTION SHIFT, ICLR 2022
